File: app/src/main/java/dev/bee/kanjianki/data/LocalStoreSimilarKanji.java
package dev.bee.kanjianki.data;

import dev.bee.kanjianki.core.RecordsImportModels;
import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import dev.bee.kanjianki.core.SimilarKanjiChoicePlanner;
import dev.bee.kanjianki.core.SimilarKanjiIndex;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Set;

abstract class LocalStoreSimilarKanji extends LocalStoreStudy {
    LocalStoreSimilarKanji(Context context) {
        super(context);
    }

    public void rebuildSimilarKanjiPairs(SimilarKanjiIndex similarIndex, long nowMillis) {
        if (similarIndex == null) {
            return;
        }
        SQLiteDatabase db = getWritableDatabase();
        db.beginTransaction();
        try {
            rebuildSimilarKanjiPairs(db, similarIndex, nowMillis);
            rebuildSimilarKanjiChoiceStates(db, nowMillis);
            db.setTransactionSuccessful();
        } finally {
            db.endTransaction();
        }
    }

    public List<RecordsImportModels.SimilarKanjiPair> allLocalSimilarPairs() {
        SQLiteDatabase db = getReadableDatabase();
        List<RecordsImportModels.SimilarKanjiPair> out = new ArrayList<>();
        try (Cursor cursor = db.query(TABLE_SIMILAR_KANJI_PAIRS, null, null, null, null, null, ORDER_SIMILAR_PAIR)) {
            while (cursor.moveToNext()) {
                out.add(readSimilarPair(cursor));
            }
        }
        return out;
    }

    public List<RecordsImportModels.SimilarKanjiPair> similarPairsForKanji(String kanji) {
        String normalized = normalizeSingleKanji(kanji);
        if (normalized.isEmpty()) {
            return Collections.emptyList();
        }
        SQLiteDatabase db = getReadableDatabase();
        List<RecordsImportModels.SimilarKanjiPair> out = new ArrayList<>();
        try (Cursor cursor = db.query(
                TABLE_SIMILAR_KANJI_PAIRS,
                null,
                "kanji_a=? OR kanji_b=?",
                new String[]{normalized, normalized},
                null,
                null,
                ORDER_SIMILAR_PAIR
        )) {
            while (cursor.moveToNext()) {
                out.add(readSimilarPair(cursor));
            }
        }
        return out;
    }

    public boolean hasSimilarLocalPair(String first, String second) {
        String kanjiA = normalizeSingleKanji(first);
        String kanjiB = normalizeSingleKanji(second);
        if (kanjiA.isEmpty() || kanjiB.isEmpty() || kanjiA.equals(kanjiB)) {
            return false;
        }
        String[] pair = canonicalSimilarPair(kanjiA, kanjiB);
        try (Cursor cursor = getReadableDatabase().query(
                TABLE_SIMILAR_KANJI_PAIRS,
                new String[]{COLUMN_KANJI_A},
                "kanji_a=? AND kanji_b=?",
                pair,
                null,
                null,
                null,
                "1"
        )) {
            return cursor.moveToFirst();
        }
    }

    public List<RecordsImportModels.SimilarKanjiChoiceCard> allSimilarChoiceCards() {
        SQLiteDatabase db = getReadableDatabase();
        List<RecordsImportModels.SimilarKanjiChoiceCard> out = new ArrayList<>();
        try (Cursor cursor = db.query(
                TABLE_SIMILAR_KANJI_CHOICE_STATE,
                null,
                null,
                null,
                null,
                null,
                "target_kanji ASC, choice_signature ASC"
        )) {
            while (cursor.moveToNext()) {
                out.add(readSimilarChoiceCard(cursor));
            }
        }
        return out;
    }

    public RecordsImportModels.SimilarKanjiChoiceCard dueSimilarChoiceForActiveTarget(String kanji, long nowMillis) {
        String target = normalizeSingleKanji(kanji);
        if (target.isEmpty()) {
            return null;
        }
        SQLiteDatabase db = getReadableDatabase();
        try (Cursor cursor = db.query(
                TABLE_SIMILAR_KANJI_CHOICE_STATE,
                null,
                "target_kanji=? AND passed_at=0 AND due_at<=?",
                new String[]{target, Long.toString(nowMillis)},
                null,
                null,
                "due_at ASC, first_seen_at ASC",
                "1"
        )) {
            if (!cursor.moveToFirst()) {
                return null;
            }
            RecordsImportModels.SimilarKanjiChoiceCard card = readSimilarChoiceCard(cursor);
            return hasPendingSimilarRepairs(db, card.targetKanji, card.choiceSignature) ? null : card;
        }
    }

    public RecordsImportModels.SimilarKanjiChoiceCard nextDueInventorySimilarChoice(Set<String> activeTargets, long nowMillis) {
        SQLiteDatabase db = getReadableDatabase();
        try (Cursor cursor = db.query(
                TABLE_SIMILAR_KANJI_CHOICE_STATE,
                null,
                "passed_at=0 AND due_at<=?",
                new String[]{Long.toString(nowMillis)},
                null,
                null,
                "due_at ASC, last_reviewed_at ASC, target_kanji ASC"
        )) {
            while (cursor.moveToNext()) {
                RecordsImportModels.SimilarKanjiChoiceCard card = readSimilarChoiceCard(cursor);
                if (activeTargets != null && activeTargets.contains(card.targetKanji)) {
                    continue;
                }
                if (!hasPendingSimilarRepairs(db, card.targetKanji, card.choiceSignature)) {
                    return card;
                }
            }
            return null;
        }
    }

    public int dueSimilarStudyTaskCount(long nowMillis) {
        return dueSimilarChoiceTaskCount(nowMillis) + dueSimilarWritingRepairTaskCount(nowMillis);
    }

    public int dueSimilarChoiceTaskCount(long nowMillis) {
        SQLiteDatabase db = getReadableDatabase();
        int count = 0;
        try (Cursor cursor = db.query(
                TABLE_SIMILAR_KANJI_CHOICE_STATE,
                new String[]{COLUMN_TARGET_KANJI, COLUMN_CHOICE_SIGNATURE},
                "passed_at=0 AND due_at<=?",
                new String[]{Long.toString(nowMillis)},
                null,
                null,
                null
        )) {
            while (cursor.moveToNext()) {
                String targetKanji = string(cursor, COLUMN_TARGET_KANJI);
                String choiceSignature = string(cursor, COLUMN_CHOICE_SIGNATURE);
                if (!hasPendingSimilarRepairs(db, targetKanji, choiceSignature)) {
                    count++;
                }
            }
        }
        return count;
    }

    public int dueSimilarWritingRepairTaskCount(long nowMillis) {
        SQLiteDatabase db = getReadableDatabase();
        try (Cursor cursor = db.rawQuery(
                "SELECT COUNT(*) FROM " + TABLE_SIMILAR_KANJI_REPAIR_QUEUE + " WHERE status=? AND due_at<=?",
                new String[]{STATUS_PENDING, Long.toString(nowMillis)}
        )) {
            cursor.moveToFirst();
            return cursor.getInt(0);
        }
    }

    public RecordsImportModels.SimilarKanjiChoiceResult submitSimilarChoice(
            RecordsImportModels.SimilarKanjiChoiceCard submitted,
            String selectedKanji,
            long nowMillis
    ) {
        return submitSimilarChoice(submitted, selectedKanji, nowMillis, true);
    }

    public RecordsImportModels.SimilarKanjiChoiceResult submitSimilarChoice(
            RecordsImportModels.SimilarKanjiChoiceCard submitted,
            String selectedKanji,
            long nowMillis,
            boolean enqueueRepairs
    ) {
        if (submitted == null) {
            return new RecordsImportModels.SimilarKanjiChoiceResult(null, selectedKanji, false, Collections.emptyList());
        }
        SQLiteDatabase db = getWritableDatabase();
        db.beginTransaction();
        try {
            RecordsImportModels.SimilarKanjiChoiceCard card = similarChoiceCard(db, submitted.targetKanji, submitted.choiceSignature);
            if (card == null) {
                card = submitted;
            }
            SimilarKanjiChoicePlanner planner = new SimilarKanjiChoicePlanner();
            RecordsImportModels.SimilarKanjiChoiceResult result = planner.evaluateSelection(card, normalizeSingleKanji(selectedKanji));

            ContentValues values = new ContentValues();
            values.put(COLUMN_LAST_REVIEWED_AT, nowMillis);
            if (result.correct) {
                values.put(COLUMN_PASSED_AT, nowMillis);
                values.put(COLUMN_CORRECT_COUNT, card.correctCount + 1);
            } else {
                values.put(COLUMN_PASSED_AT, 0L);
                values.put(COLUMN_DUE_AT, nowMillis);
                values.put(COLUMN_WRONG_COUNT, card.wrongCount + 1);
            }
            db.update(
                    TABLE_SIMILAR_KANJI_CHOICE_STATE,
                    values,
                    WHERE_SIMILAR_CHOICE,
                    new String[]{card.targetKanji, card.choiceSignature}
            );

            ContentValues log = new ContentValues();
            log.put(COLUMN_TARGET_KANJI, card.targetKanji);
            log.put(COLUMN_CHOICE_SIGNATURE, card.choiceSignature);
            log.put("selected_kanji", result.selectedKanji);
            log.put("correct", result.correct ? 1 : 0);
            log.put(COLUMN_REVIEWED_AT, nowMillis);
            db.insert(TABLE_SIMILAR_KANJI_REVIEW_LOG, null, log);

            if (!result.correct && enqueueRepairs) {
                for (String repairKanji : result.repairKanji) {
                    enqueueSimilarWritingRepair(db, card, repairKanji, result.selectedKanji, nowMillis);
                }
            }
            db.setTransactionSuccessful();
            return result;
        } finally {
            db.endTransaction();
        }
    }

    public RecordsImportModels.SimilarKanjiWritingRepair nextDueSimilarWritingRepair(long nowMillis) {
        List<RecordsImportModels.SimilarKanjiWritingRepair> repairs = dueSimilarWritingRepairs(nowMillis);
        return repairs.isEmpty() ? null : repairs.get(0);
    }

    public List<RecordsImportModels.SimilarKanjiWritingRepair> dueSimilarWritingRepairs(long nowMillis) {
        SQLiteDatabase db = getReadableDatabase();
        List<RecordsImportModels.SimilarKanjiWritingRepair> out = new ArrayList<>();
        try (Cursor cursor = db.query(
                TABLE_SIMILAR_KANJI_REPAIR_QUEUE,
                null,
                "status=? AND due_at<=?",
                new String[]{STATUS_PENDING, Long.toString(nowMillis)},
                null,
                null,
                "created_at ASC, id ASC",
                null
        )) {
            while (cursor.moveToNext()) {
                out.add(readSimilarWritingRepair(cursor));
            }
        }
        return out;
    }

    public void saveSimilarWritingRepair(RecordsImportModels.SimilarKanjiWritingRepair repair) {
        if (repair == null || repair.id <= 0L) {
            return;
        }
        ContentValues values = new ContentValues();
        values.put(COLUMN_ACTIVE_TOKEN, repair.activeToken);
        values.put(COLUMN_UPDATED_AT, repair.updatedAtMillis);
        getWritableDatabase().update(
                TABLE_SIMILAR_KANJI_REPAIR_QUEUE,
                values,
                "id=? AND status=?",
                new String[]{Long.toString(repair.id), STATUS_PENDING}
        );
    }

    public boolean finishSimilarWritingRepair(long repairId, String token, boolean passed, long nowMillis) {
        SQLiteDatabase db = getWritableDatabase();
        db.beginTransaction();
        try {
            RecordsImportModels.SimilarKanjiWritingRepair current = similarWritingRepair(db, repairId);
            if (current == null || !STATUS_PENDING.equals(current.status)) {
                return false;
            }
            if (!current.activeToken.isEmpty() && !current.activeToken.equals(token == null ? "" : token)) {
                return false;
            }
            ContentValues values = new ContentValues();
            values.put(COLUMN_ACTIVE_TOKEN, "");
            values.put(COLUMN_UPDATED_AT, nowMillis);
            if (passed) {
                values.put(COLUMN_STATUS, STATUS_COMPLETE);
                values.put(COLUMN_COMPLETED_AT, nowMillis);
            } else {
                values.put(COLUMN_ATTEMPTS, current.attempts + 1);
                values.put(COLUMN_DUE_AT, nowMillis);
            }
            db.update(TABLE_SIMILAR_KANJI_REPAIR_QUEUE, values, "id=?", new String[]{Long.toString(repairId)});
            db.setTransactionSuccessful();
            return true;
        } finally {
            db.endTransaction();
        }
    }
}

File: app/src/main/kotlin/dev/bee/kanjianki/data/LocalStoreSimilarKanji.kt
package dev.bee.kanjianki.data

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import androidx.core.database.sqlite.transaction
import dev.bee.kanjianki.core.ConfusionPairMiner
import dev.bee.kanjianki.core.RecordsBase
import dev.bee.kanjianki.core.RecordsImportModels
import dev.bee.kanjianki.core.SimilarKanjiChoicePlanner
import dev.bee.kanjianki.core.SimilarKanjiChoiceReviewPolicy
import dev.bee.kanjianki.core.SimilarKanjiIndex
import dev.bee.kanjianki.core.SimilarKanjiRepairPolicy

internal abstract class LocalStoreSimilarKanji(context: Context?) : LocalStoreStudy(context) {
    fun rebuildSimilarKanjiPairs(similarIndex: SimilarKanjiIndex?, nowMillis: Long) {
        if (similarIndex == null) {
            return
        }
        writableDatabase.transaction {
            rebuildSimilarKanjiPairs(this, similarIndex, nowMillis)
            rebuildSimilarKanjiChoiceStates(this, nowMillis)
        }
        clearStudyItemsCache()
    }

    fun allLocalSimilarPairs(): List<RecordsImportModels.SimilarKanjiPair> {
        val out = ArrayList<RecordsImportModels.SimilarKanjiPair>()
        readableDatabase.query(TABLE_SIMILAR_KANJI_PAIRS, null, null, null, null, null, ORDER_SIMILAR_PAIR).use {
            while (it.moveToNext()) {
                out.add(readSimilarPair(it))
            }
        }
        return out
    }

    fun similarPairsForKanji(kanji: String?): List<RecordsImportModels.SimilarKanjiPair> {
        val normalized = normalizeSingleKanji(kanji)
        if (normalized.isEmpty()) {
            return emptyList()
        }
        val out = ArrayList<RecordsImportModels.SimilarKanjiPair>()
        readableDatabase.query(
            TABLE_SIMILAR_KANJI_PAIRS,
            null,
            "kanji_a=? OR kanji_b=?",
            arrayOf(normalized, normalized),
            null,
            null,
            ORDER_SIMILAR_PAIR
        ).use {
            while (it.moveToNext()) {
                out.add(readSimilarPair(it))
            }
        }
        return out
    }

    fun hasSimilarLocalPair(first: String?, second: String?): Boolean {
        val kanjiA = normalizeSingleKanji(first)
        val kanjiB = normalizeSingleKanji(second)
        if (kanjiA.isEmpty() || kanjiB.isEmpty() || kanjiA == kanjiB) {
            return false
        }
        val pair = canonicalSimilarPair(kanjiA, kanjiB)
        readableDatabase.query(
            TABLE_SIMILAR_KANJI_PAIRS,
            arrayOf(COLUMN_KANJI_A),
            "kanji_a=? AND kanji_b=?",
            pair,
            null,
            null,
            null,
            "1"
        ).use {
            return it.moveToFirst()
        }
    }

    /**
     * Aggregated wrong-pick counts (target -> selected -> count) from the
     * choice review log, limited to the confusion-mining window. Used to
     * weight multiple-choice distractors toward historically confused kanji.
     */
    fun choiceWrongPickCounts(nowMillis: Long): Map<String, Map<String, Int>> {
        return choiceWrongPickCounts(readableDatabase, ConfusionPairMiner.windowStartMillis(nowMillis))
    }

    fun allSimilarChoiceCards(): List<RecordsImportModels.SimilarKanjiChoiceCard> {
        val out = ArrayList<RecordsImportModels.SimilarKanjiChoiceCard>()
        readableDatabase.query(
            TABLE_SIMILAR_KANJI_CHOICE_STATE,
            null,
            null,
            null,
            null,
            null,
            "target_kanji ASC, choice_signature ASC"
        ).use {
            while (it.moveToNext()) {
                out.add(readSimilarChoiceCard(it))
            }
        }
        return out
    }

    fun dueSimilarChoiceForActiveTarget(
        kanji: String?,
        nowMillis: Long,
    ): RecordsImportModels.SimilarKanjiChoiceCard? {
        val target = normalizeSingleKanji(kanji)
        if (target.isEmpty()) {
            return null
        }
        val db = readableDatabase
        db.query(
            TABLE_SIMILAR_KANJI_CHOICE_STATE,
            null,
            "target_kanji=? AND passed_at=0 AND due_at<=?",
            arrayOf(target, nowMillis.toString()),
            null,
            null,
            "due_at ASC, first_seen_at ASC",
            "1"
        ).use {
            if (!it.moveToFirst()) {
                return null
            }
            val card = readSimilarChoiceCard(it)
            return if (hasPendingSimilarRepairs(db, card.targetKanji, card.choiceSignature)) null else card
        }
    }

    fun nextDueInventorySimilarChoice(
        activeTargets: Set<String>?,
        nowMillis: Long,
    ): RecordsImportModels.SimilarKanjiChoiceCard? {
        val db = readableDatabase
        db.query(
            TABLE_SIMILAR_KANJI_CHOICE_STATE,
            null,
            "passed_at=0 AND due_at<=?",
            arrayOf(nowMillis.toString()),
            null,
            null,
            "due_at ASC, last_reviewed_at ASC, target_kanji ASC"
        ).use {
            while (it.moveToNext()) {
                val card = readSimilarChoiceCard(it)
                if (activeTargets != null && activeTargets.contains(card.targetKanji)) {
                    continue
                }
                if (!hasPendingSimilarRepairs(db, card.targetKanji, card.choiceSignature)) {
                    return card
                }
            }
            return null
        }
    }

    fun dueSimilarStudyTaskCount(nowMillis: Long): Int {
        return dueSimilarChoiceTaskCount(nowMillis) + dueSimilarWritingRepairTaskCount(nowMillis)
    }

    fun dueSimilarChoiceTaskCount(nowMillis: Long): Int {
        val db = readableDatabase
        var count = 0
        db.query(
            TABLE_SIMILAR_KANJI_CHOICE_STATE,
            arrayOf(COLUMN_TARGET_KANJI, COLUMN_CHOICE_SIGNATURE),
            "passed_at=0 AND due_at<=?",
            arrayOf(nowMillis.toString()),
            null,
            null,
            null
        ).use {
            while (it.moveToNext()) {
                val targetKanji = string(it, COLUMN_TARGET_KANJI)
                val choiceSignature = string(it, COLUMN_CHOICE_SIGNATURE)
                if (!hasPendingSimilarRepairs(db, targetKanji, choiceSignature)) {
                    count++
                }
            }
        }
        return count
    }

    fun dueSimilarWritingRepairTaskCount(nowMillis: Long): Int {
        readableDatabase.rawQuery(
            "SELECT COUNT(*) FROM $TABLE_SIMILAR_KANJI_REPAIR_QUEUE WHERE status=? AND due_at<=?",
            arrayOf(STATUS_PENDING, nowMillis.toString())
        ).use {
            it.moveToFirst()
            return it.getInt(0)
        }
    }

    fun submitSimilarChoice(
        submitted: RecordsImportModels.SimilarKanjiChoiceCard?,
        selectedKanji: String?,
        nowMillis: Long,
    ): RecordsImportModels.SimilarKanjiChoiceResult {
        return submitSimilarChoice(submitted, selectedKanji, nowMillis, true)
    }

    /**
     * Records a multiple-choice pick for confusion-pair mining. Used by rungs
     * that render a choice grid outside the similar-kanji practice tables
     * (currently meaning_kanji); the similar_kanji path logs inline during
     * submitSimilarChoice.
     */
    fun recordChoiceReviewLog(
        targetKanji: String?,
        choiceSignature: String?,
        selectedKanji: String?,
        correct: Boolean,
        rung: String?,
        nowMillis: Long,
    ) {
        val target = normalizeSingleKanji(targetKanji)
        val selected = normalizeSingleKanji(selectedKanji)
        if (target.isEmpty() || selected.isEmpty()) {
            return
        }
        val log = ContentValues()
        log.put(COLUMN_TARGET_KANJI, target)
        log.put(COLUMN_CHOICE_SIGNATURE, choiceSignature ?: "")
        log.put("selected_kanji", selected)
        log.put("correct", if (correct) 1 else 0)
        log.put(COLUMN_REVIEWED_AT, nowMillis)
        log.put(COLUMN_RUNG, rung?.trim().takeUnless { it.isNullOrEmpty() } ?: RecordsBase.LadderRung.SIMILAR_KANJI.wireName())
        writableDatabase.insert(TABLE_SIMILAR_KANJI_REVIEW_LOG, null, log)
    }

    fun submitSimilarChoice(
        submitted: RecordsImportModels.SimilarKanjiChoiceCard?,
        selectedKanji: String?,
        nowMillis: Long,
        enqueueRepairs: Boolean,
    ): RecordsImportModels.SimilarKanjiChoiceResult {
        if (submitted == null) {
            return RecordsImportModels.SimilarKanjiChoiceResult(null, selectedKanji, false, emptyList())
        }
        var result: RecordsImportModels.SimilarKanjiChoiceResult? = null
        writableDatabase.transaction {
            var card = similarChoiceCard(this, submitted.targetKanji, submitted.choiceSignature)
            if (card == null) {
                card = submitted
            }
            val planner = SimilarKanjiChoicePlanner()
            val evaluated = planner.evaluateSelection(card, normalizeSingleKanji(selectedKanji))
            val reviewUpdate = SimilarKanjiChoiceReviewPolicy.reviewUpdate(card, evaluated, nowMillis)

            val values = ContentValues()
            values.put(COLUMN_LAST_REVIEWED_AT, reviewUpdate.lastReviewedAtMillis())
            values.put(COLUMN_PASSED_AT, reviewUpdate.passedAtMillis())
            reviewUpdate.dueAtMillis()?.let { values.put(COLUMN_DUE_AT, it) }
            reviewUpdate.correctCount()?.let { values.put(COLUMN_CORRECT_COUNT, it) }
            reviewUpdate.wrongCount()?.let { values.put(COLUMN_WRONG_COUNT, it) }
            update(
                TABLE_SIMILAR_KANJI_CHOICE_STATE,
                values,
                WHERE_SIMILAR_CHOICE,
                arrayOf(card.targetKanji, card.choiceSignature)
            )

            val log = ContentValues()
            log.put(COLUMN_TARGET_KANJI, card.targetKanji)
            log.put(COLUMN_CHOICE_SIGNATURE, card.choiceSignature)
            log.put("selected_kanji", evaluated.selectedKanji)
            log.put("correct", if (evaluated.correct) 1 else 0)
            log.put(COLUMN_REVIEWED_AT, nowMillis)
            log.put(COLUMN_RUNG, RecordsBase.LadderRung.SIMILAR_KANJI.wireName())
            insert(TABLE_SIMILAR_KANJI_REVIEW_LOG, null, log)

            if (!evaluated.correct && enqueueRepairs) {
                for (repairKanji in evaluated.repairKanji) {
                    enqueueSimilarWritingRepair(this, card, repairKanji, evaluated.selectedKanji, nowMillis)
                }
            }
            result = evaluated
        }
        return result ?: RecordsImportModels.SimilarKanjiChoiceResult(null, selectedKanji, false, emptyList())
    }

    fun nextDueSimilarWritingRepair(nowMillis: Long): RecordsImportModels.SimilarKanjiWritingRepair? {
        val repairs = dueSimilarWritingRepairs(nowMillis)
        return if (repairs.isEmpty()) null else repairs[0]
    }

    fun dueSimilarWritingRepairs(nowMillis: Long): List<RecordsImportModels.SimilarKanjiWritingRepair> {
        val out = ArrayList<RecordsImportModels.SimilarKanjiWritingRepair>()
        readableDatabase.query(
            TABLE_SIMILAR_KANJI_REPAIR_QUEUE,
            null,
            "status=? AND due_at<=?",
            arrayOf(STATUS_PENDING, nowMillis.toString()),
            null,
            null,
            "created_at ASC, id ASC",
            null
        ).use {
            while (it.moveToNext()) {
                out.add(readSimilarWritingRepair(it))
            }
        }
        return out
    }

    fun saveSimilarWritingRepair(repair: RecordsImportModels.SimilarKanjiWritingRepair?) {
        if (repair == null || repair.id <= 0L) {
            return
        }
        val values = ContentValues()
        values.put(COLUMN_ACTIVE_TOKEN, repair.activeToken)
        values.put(COLUMN_UPDATED_AT, repair.updatedAtMillis)
        writableDatabase.update(
            TABLE_SIMILAR_KANJI_REPAIR_QUEUE,
            values,
            "id=? AND status=?",
            arrayOf(repair.id.toString(), STATUS_PENDING)
        )
    }

    fun finishSimilarWritingRepair(
        repairId: Long,
        token: String?,
        passed: Boolean,
        nowMillis: Long,
    ): Boolean {
        var finished = false
        writableDatabase.transaction {
            val current = similarWritingRepair(this, repairId)
            if (current == null || current.status != STATUS_PENDING) {
                return@transaction
            }
            if (current.activeToken.isNotEmpty() && current.activeToken != (token ?: "")) {
                return@transaction
            }
            val finishUpdate = SimilarKanjiRepairPolicy.finishUpdate(current, passed, nowMillis)
            val values = ContentValues()
            values.put(COLUMN_ACTIVE_TOKEN, finishUpdate.activeToken())
            values.put(COLUMN_UPDATED_AT, finishUpdate.updatedAtMillis())
            finishUpdate.status()?.let { values.put(COLUMN_STATUS, it) }
            finishUpdate.completedAtMillis()?.let { values.put(COLUMN_COMPLETED_AT, it) }
            finishUpdate.attempts()?.let { values.put(COLUMN_ATTEMPTS, it) }
            finishUpdate.dueAtMillis()?.let { values.put(COLUMN_DUE_AT, it) }
            update(TABLE_SIMILAR_KANJI_REPAIR_QUEUE, values, "id=?", arrayOf(repairId.toString()))
            finished = true
        }
        return finished
    }

    fun skipSimilarWritingRepair(
        repairId: Long,
        token: String?,
        nowMillis: Long,
    ): Boolean {
        var finished = false
        writableDatabase.transaction {
            val current = similarWritingRepair(this, repairId)
            if (current == null || current.status != STATUS_PENDING) {
                return@transaction
            }
            if (current.activeToken.isNotEmpty() && current.activeToken != (token ?: "")) {
                return@transaction
            }
            val finishUpdate = SimilarKanjiRepairPolicy.skipUpdate(current, nowMillis)
            val values = ContentValues()
            values.put(COLUMN_ACTIVE_TOKEN, finishUpdate.activeToken())
            values.put(COLUMN_UPDATED_AT, finishUpdate.updatedAtMillis())
            finishUpdate.status()?.let { values.put(COLUMN_STATUS, it) }
            finishUpdate.completedAtMillis()?.let { values.put(COLUMN_COMPLETED_AT, it) }
            finishUpdate.attempts()?.let { values.put(COLUMN_ATTEMPTS, it) }
            finishUpdate.dueAtMillis()?.let { values.put(COLUMN_DUE_AT, it) }
            update(TABLE_SIMILAR_KANJI_REPAIR_QUEUE, values, "id=?", arrayOf(repairId.toString()))
            finished = true
        }
        return finished
    }
}
